game-core/src/main/java/games/strategy/engine/data/util/BreadthFirstSearch.java

```diff
@@ -0,0 +1,91 @@
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.ObjectUtils;
+
+/**
+ * Implements Breadth First Search (BFS) to traverse / find territories. Since the search criteria
+ * varies depending on the use case, the class is designed to be sub-classed, with methods visit()
+ * and shouldContinueSearch() that can be overridden to customize the behavior.
+ */
+public class BreadthFirstSearch {
+  private final GameMap map;
+  @Getter private final Set<Territory> visited;
+  private final ArrayDeque<Territory> territoriesToCheck;
+  private final Predicate<Territory> cond;
+
+  public BreadthFirstSearch(final Territory startTerritory, final Predicate<Territory> cond) {
+    this.map = startTerritory.getData().getMap();
+    this.visited = new HashSet<Territory>(List.of(startTerritory));
+    this.territoriesToCheck = new ArrayDeque<Territory>(List.of(startTerritory));
+    this.cond = cond;
+  }
+
+  public BreadthFirstSearch(final Territory startTerritory) {
+    this(startTerritory, Matches.always());
+  }
+
+  /**
+   * Called when a new territory is encountered. Can be overridden to provide custom search
+   * behavior.
+   *
+   * @param territory The new territory.
+   */
+  public void visit(final Territory territory) {}
+
+  /**
+   * Called after all territories within the specified distance have been searched. Can be
+   * overridden to terminate the search.
+   *
+   * @param distanceSearched The current distance searched
+   * @return Whether the search should continue.
+   */
+  public boolean shouldContinueSearch(final int distanceSearched) {
+    return true;
+  }
+
+  /**
+   * Performs the search. It will end when either all territories have been visited or
+   * shouldContinueSearch() returns false.
+   */
+  public void search() {
+    // Since we process territories in order of distance, we can keep track of the last territory
+    // at the current distance that's in the territoriesToCheck queue. When we encounter it, we
+    // increment the distance and update lastTerritoryAtCurrentDistance.
+    int currentDistance = 0;
+    Territory lastTerritoryAtCurrentDistance = territoriesToCheck.peekLast();
+    while (!territoriesToCheck.isEmpty()) {
+      final Territory territory = checkNextTerritory();
+
+      // If we just processed the last territory at the current distance, increment the distance
+      // and set the territory at which we need to update it again to be the last one added.
+      if (ObjectUtils.referenceEquals(territory, lastTerritoryAtCurrentDistance)) {
+        currentDistance++;
+        if (!shouldContinueSearch(currentDistance)) {
+          return;
+        }
+        lastTerritoryAtCurrentDistance = territoriesToCheck.peekLast();
+      }
+    }
+  }
+
+  private Territory checkNextTerritory() {
+    final Territory territory = territoriesToCheck.removeFirst();
+    // Note: We don't pass cond to getNeighbors() because that implementation is much slower.
```

Comment:
Perhaps worth clarifying:
The reason the implementation is slower is because it tries to parallelize checking neighbours, which adds too much overhead for the ususally small amount of neighbours + it actually copies the collection instead of just returning it.

I'd suggest to clarify that a bit why it is much slower in this comment.
Maybe you can even drop the parallelism entirely, I'm not really convinced it actually provides a speed boost elsewhere.
I introduced this change in #4077 (apparently) because it yielded some performance gains for route-drawing, but I haven't really tested anything else as far as I can remember.

The `GameMap` class is a mess (I'm currently looking at it) and uses way too much recursion in combination with code constructs that aren't really supposed to be run recursively like this (Streams that are collected and then re-streamed in another layer add way too much overhead)

Reply:
I haven't dug into why it's slower. I just measured things after my initial refactoring that did this and it was slower. I think we can rewrite that later, but didn't want to expand the scope of this change.

Reply:
No worries, just wanted to provide some extra information for documentation purposes.